### PR TITLE
Make explainer work with categorical targets not present in the training data

### LINKED ISSUE #489: Nonconfrmist issue when test dataset doesn't contain values that existed in the train dataset
When there's a categorical target and nonconformist isn't getting a certain categorical value present in the train set in the validation dataset this seems to lead to errors:

```
    self.model_analysis, self.runtime_analyzer = model_analyzer(data=test_data,train_data=train_data,disable_column_importance=False,fixed_significance=None,confidence_normalizer=False,stats_info=self.statistical_analysis,ts_cfg=self.problem_definition.timeseries_settings,accuracy_functions=self.accuracy_functions,predictor=self.ensemble,target=self.target,dtype_dict=self.dtype_dict,positive_domain=self.statistical_analysis.positive_domain)
  File "/home/george/lightwood/lightwood/analysis/model_analyzer.py", line 153, in model_analyzer
    icp_df, y = clean_df(icp_df, target, is_classification, runtime_analyzer.get('label_encoders', None))
  File "/home/george/lightwood/lightwood/analysis/nc/util.py", line 15, in clean_df
    y = np.array([cats.index(i) for i in y])
  File "/home/george/lightwood/lightwood/analysis/nc/util.py", line 15, in <listcomp>
    y = np.array([cats.index(i) for i in y])
ValueError: 'IT' is not in list
```

Was fixed in native so hopefully we can just copy-paste from there (?)

### PR DESCRIPTION
**Pull request by George3d6**
## Why 

NC crashes if there's a categorical target value in the validation/test data (the one that it's fit on) which it is not present in the train data.

## How

We fix this by just adding an `if this is a value not present in the train data distribution for the target, replace it with the "unknown category" label`

Added a test to catch and replicate this behaviour 

## Other

Added type signatures to a bunch of the explainer modules

closes #489  [Note: the description in the issue is bad because initially I thought the bug was smth else, but then I couldn't replace]

### CODE REVIEW

**Comment by paxcema** [comment]
There's lots of shared logic with #497 in here, so I'll wait until that gets merged to staging (and then to this one) before reviewing.

**Review by hakunanatasha — COMMENTED** [review]
If Patricio approves of the NC logic I am ok with this.

File: tests/integration/basic/test_weird_target_dist.py
Comment (btseytlin) [inline tests/integration/basic/test_weird_target_dist.py:20]: How about some assert? Maybe assert that the missing targets were encoded as expected?

Comment (George3d6) [reply tests/integration/basic/test_weird_target_dist.py:20]: No point, we just want to make sure it doesn't crash when the validation see contains things that aren't in train, that's all. And that is implicitly happening due to the number of distinct targets we generate. Might be worth explaining this better though...

Comment (paxcema) [reply tests/integration/basic/test_weird_target_dist.py:30]: Even if it is implicitly happening (which imo it's not very easy to see unless you're familiar with lightwood internals) I think it's better to have 2 different dataframes here to guarantee there's stuff the ICP hasn't seen in training when doing the `predict()` call.

File: tests/unit_tests/analysis/test_nc_norm.py
Comment (paxcema) [inline tests/unit_tests/analysis/test_nc_norm.py:13]: Remove `print()` call

File: tests/unit_tests/analysis/test_nc_norm.py
Comment (paxcema) [inline tests/unit_tests/analysis/test_nc_norm.py:24]: Remove `print()` call

File: lightwood/analysis/nc/util.py
Comment (paxcema) [inline lightwood/analysis/nc/util.py:15]: Typo, should read `__mdb_unknown_cat`

File: lightwood/analysis/nc/icp.py
Comment (paxcema) [inline lightwood/analysis/nc/icp.py:111]: Missing type hints, same as below would be OK

File: lightwood/analysis/nc/icp.py
Comment (paxcema) [inline lightwood/analysis/nc/icp.py:104]: I believe the return hint here should be `None`

**Review by paxcema — CHANGES_REQUESTED** [review]

File: tests/integration/basic/test_weird_target_dist.py
Comment (George3d6) [reply tests/integration/basic/test_weird_target_dist.py:30]: Then I'd have to call the training functions one by one instead. But in practice you can test the test works (as I did) by running it on staging, and seeing it fails with the same error listed in the issue

**Review by paxcema — APPROVED** [review]

**Pull request merged**
